### ISSUE TITLE
CTC部分loss降到30左右降不下去了

### ISSUE THREAD

**Issue by yjiangling** [YimengZhu/ChineseASR#1]
博主你好，我也遇到了跟你相似的问题，联合训练过程中发现CTC部分的loss降到30左右降不下去了，请问你解决了这个问题吗？

**Comment by YimengZhu** [comment]
你说的哪个模型啊，最后我loss都降下去了，主要是靠调learning rate。

另外你是在哪个地方看到这个repo的啊，我没记得我写过啥博客啊。。。

**Comment by yjiangling** [comment]
> 你说的哪个模型啊，最后我loss都降下去了，主要是靠调learning rate。
> 
> 另外你是在哪个地方看到这个repo的啊，我没记得我写过啥博客啊。。。

噢，我是在知乎上看的一个问题 https://www.zhihu.com/question/401053099 可能是用PyTorch复现你的代码吧。我最近也在做一个CTC+Transformer的ASR模型，但是很奇怪的是CTC这部分loss每次降到30左右就非常难降下去，我尝试了直接在几层MultiHeadAttention后面接上CTC loss，这样单独训练一个CTC模型发现loss也是降不下去，请问博主你的学习率设置的是多少啊？CTC前面你用了几层Attention啊？

**Comment by YimengZhu** [comment]
I didn't manage to install Chinese input method on the laptop I'm currently using so I'll answer your question in English.

1. Are you sure you are using **transformer**, or otherwise, **solely attention layers** with CTC? In my opinion, you don't really need CTC loss with transformer as it has the encoder-decoder architecture to accomplish the alignment between input and output sequences automatically, which is originally supposed to be done by CTC. In my case, I was actually implementing [this paper](https://arxiv.org/pdf/1901.10055.pdf). Note that in the paper, it is not the Transformer but a bunch of stacked self attention layers with a CTC layer upon them.

2. In my case, the hyper parameter are essentially the same as in the original paper. I cannot remember the correct number of layers but the total parameter number is approx. 35M. I guess it was between 13~15 self attention layers. But you should not simply try mines as this number need to be inferred from you dev set result. I'm not sure my hyper parameter setting works for you as well because I don't know the dataset you have. If you have a small dataset with e.g. 200 hours you can try mines. But if your set is massive and cross domain, my hyper parameter setting will definitely screw up in your training.

3. With regarding to learning rate, you should choose the optimizer I implemented (with warm up and decreasing phases). It is also very important to try with different the parameters `k` and `warmup_step`. This can significantly affect the loss in the training.

**Comment by yjiangling** [comment]
> I didn't manage to install Chinese input method on the laptop I'm currently using so I'll answer your question in English.
> 
> 1. Are you sure you are using **transformer**, or otherwise, **solely attention layers** with CTC? In my opinion, you don't really need CTC loss with transformer as it has the encoder-decoder architecture to accomplish the alignment between input and output sequences automatically, which is originally supposed to be done by CTC. In my case, I was actually implementing [this paper](https://arxiv.org/pdf/1901.10055.pdf). Note that in the paper, it is not the Transformer but a bunch of stacked self attention layers with a CTC layer upon them.
> 2. In my case, the hyper parameter are essentially the same as in the original paper. I cannot remember the correct number of layers but the total parameter number is approx. 35M. I guess it was between 13~15 self attention layers. But you should not simply try mines as this number need to be inferred from you dev set result. I'm not sure my hyper parameter setting works for you as well because I don't know the dataset you have. If you have a small dataset with e.g. 200 hours you can try mines. But if your set is massive and cross domain, my hyper parameter setting will definitely screw up in your training.
> 3. With regarding to learning rate, you should choose the optimizer I implemented (with warm up and decreasing phases). It is also very important to try with different the parameters `k` and `warmup_step`. This can significantly affect the loss in the training.

好的，谢谢，我看了你的代码，我做的Transformer+CTC是为了做流式语音识别的，CTC是为了在训练阶段做对齐的辅助工具的，便于控制Transformer模型Encoder-Decoder之间的Attention计算。由于开始loss很难降下去，后来我就直接尝试了用6层MultiHeadAttention+CTC这样的声学模型去训练，跟你的项目中SAN+CTC的模型应该是基本上一样的。不过现在loss都降下去了，好像可能是我的warmup_step参数设置不合理导致中间很多个epoch loss一直降不下来，不过到后面学习率降下来后又开始下降了，感谢博主的解答，谢谢~

**Comment by shibo0** [comment]
@yjiangling @YimengZhu 请问，完整的Transformer模型（Encoder + Decoder）能不能和CTCloss一块使用？在我的实验中，完整的Transformer和CTCLoss的效果很好，但是在测试的时候出了问题。我也搜了很多GitHub上的代码，但是大多数的loss函数用的是CrossEntroy。

**Comment by yjiangling** [comment]
> @yjiangling @YimengZhu 请问，完整的Transformer模型（Encoder + Decoder）能不能和CTCloss一块使用？在我的实验中，完整的Transformer和CTCLoss的效果很好，但是在测试的时候出了问题。我也搜了很多GitHub上的代码，但是大多数的loss函数用的是CrossEntroy。

可以，把Encoder端CTC loss和Decoder端CE loss一起训练可以得到很好的效果

**Comment by LindgeW** [comment]
请问训练到半中间ctc loss突然变成nan是什么情况？
